.github/workflows/ci.yml

```diff
@@ -0,0 +1,123 @@
+      id: ruff-check
+      run: |
+        poetry run ruff check --output-format github
+        poetry run ruff check --output-format junit --output-file junit/ruff-check-results-${{ runner.os }}-${{ matrix.python_version }}.xml
+      if: always()
+
+    - name: Run ruff format --diff
+      run: |
+        poetry run ruff format --diff
+        echo '## :microscope: ruff check errors' >> $env:GITHUB_STEP_SUMMARY
+        echo '' >> $env:GITHUB_STEP_SUMMARY
+        echo '### Lint rule errors' >> $env:GITHUB_STEP_SUMMARY
+        echo '```' >> $env:GITHUB_STEP_SUMMARY
+        poetry run ruff check --output-format grouped >> $env:GITHUB_STEP_SUMMARY
+        echo '```' >> $env:GITHUB_STEP_SUMMARY
+        echo '- :green_circle: Use `poe fix` to apply safe fixes' >> $env:GITHUB_STEP_SUMMARY
+        echo '- :warning: Use `ruff check --fix --unsafe-fixes` to apply experimental fixes' >> $env:GITHUB_STEP_SUMMARY
+        echo '- :book: Use `ruff rule Z123 | mdv -` to see examples' >> $env:GITHUB_STEP_SUMMARY
+        echo '  - `mdv` is a [terminal viewer](https://github.com/axiros/terminal_markdown_viewer) for markdown' >> $env:GITHUB_STEP_SUMMARY
+        echo '  - Full list on https://docs.astral.sh/ruff/rules/' >> $env:GITHUB_STEP_SUMMARY
+        echo '' >> $env:GITHUB_STEP_SUMMARY
+        echo '### Format errors' >> $env:GITHUB_STEP_SUMMARY
+        echo '```diff' >> $env:GITHUB_STEP_SUMMARY
+        poetry run ruff format --diff >> $env:GITHUB_STEP_SUMMARY
+        echo '```' >> $env:GITHUB_STEP_SUMMARY
+        echo ' - :green_circle: Use `ruff format` to fix errors' >> $env:GITHUB_STEP_SUMMARY
+        echo '' >> $env:GITHUB_STEP_SUMMARY
+      if: ${{ failure() && steps.ruff-check.conclusion == 'failure' }}
+      continue-on-error: true
+
+    - name: Run pyright
+      run: |
+        poetry run pyright --dependencies
+        echo '## :microscope: pyright results' >> $env:GITHUB_STEP_SUMMARY
+        echo '```' >> $env:GITHUB_STEP_SUMMARY
+        poetry run pyright --dependencies >> $env:GITHUB_STEP_SUMMARY
+        echo '```' >> $env:GITHUB_STEP_SUMMARY
+      if: always()
+
+    - name: Create result report
+      id: result-report
+      uses: dorny/test-reporter@v1.9.1
+      with:
+        name: Results for Python ${{ matrix.python_version }} on ${{ runner.os }}
+        path: junit/*.xml
+        reporter: 'java-junit'
+        fail-on-error: false
+      if: always()
+
+    - name: Get report ID
+      run: |
+        $report_id_segment = '${{steps.result-report.outputs.url}}'.split('/')[-1]
+        echo "REPORT_ID=$report_id_segment" >> $env:GITHUB_ENV
+      if: always()
+
+    - name: Build package
+      run: poetry build --verbose --format sdist
+      if: always()
+
+    - name: Upload package
+      id: upload-package
+      uses: actions/upload-artifact@v4
+      with:
+        name: qtpy_datalogger for Python ${{ matrix.python_version }} on ${{ runner.os }}
+        path: dist/*.tar.gz
+      if: always()
```

Comment:
This workflow hasn't been tested when building the package fails. When we encounter that situation, we'll use the reproduction to debug and fix it.